src/metabolic_ninja/worker/decorators.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2018, Novo Nordisk Foundation Center for Biosustainability,
+# Technical University of Denmark.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import logging
+import multiprocessing
+import os
+import sys
+
+import sentry_sdk
+
+
+logger = logging.getLogger(__name__)
+
+
+class TaskFailedException(Exception):
+    """Thrown if a task throws an unhandled exception."""
+
+    pass
+
+
+def task(function):
+    """
+    Execute the given function in a child process.
+
+    Use this as a decorator on a function to make sure it's called in a forked
+    process. The return value of the process will be passed through a pipe, and
+    so is subject to the same restrictions[1], namely, the object must be
+    pickleable and not too large (approximately 32 MiB+).
+
+    If the child process throws an exception, it will be logged, reported to Sentry, the
+    database status will be updated and `TaskFailedException` will be raised.
+
+    [1] https://docs.python.org/3/library/multiprocessing.html#multiprocessing.connection.Connection.send  # noqa
+    """
+
+    def runner(pipe, job, *args, **kwargs):
+        # This is the function called in a new process.
+        # Sentry needs to be initialized here (in addition to the main process).
+        sentry_sdk.init(dsn=os.environ.get("SENTRY_DSN"))
+        # Call the wrapped function with the given arguments and pass the return value
+        # back through a pipe.
+        try:
+            retval = function(job, *args, **kwargs)
+        except Exception as exception:
+            job.save(status="FAILURE")
+            logger.exception(exception)
+            sentry_sdk.capture_exception(exception)
+            # Wait for the sentry event to be sent; otherwise we'd exit the process too
+            # soon.
+            sentry_sdk.flush()
+            sys.exit(-1)
```

Comment:
-1 is an unusual signal. Does it have a special meaning here?

Reply:
Ah, that's out of habit. -1 in two's complement resolves to the largest possible exit code. The general idea is that for non-zero exit codes, special meanings can be assigned to the immediate numbers (1, 2, ...), and -1 is generally used for a non specific error code to avoid conflict with other numbers. We'll probably never assign meaning to exit codes, the parent process just needs to distinguish success from error here, so it's not too important - just a general pattern/habit.